rfcs/0028-nix-release-model.md

```diff
@@ -0,0 +1,158 @@
+3. The release manager will branch off a `#-maintenance` branch
+   (e.g. `2.1-maintenance`) in advance of any mainline release (that
+   is, a new release from active development).
+4. Changes aimed at getting the release ready (fixing bugs, improving
+   docs, etc.) should target the `#-maintenance` branch and then
+   be merged forward into `master`. No commit should exist on the
+   `#-maintenance` branch during the stabilization phase without very
+   soon thereafter being merged into `master` except under rare
+   circumstances. This implies that especially large changes ready
+   around a release should either be included before branch-off (with
+   extra time to stabilize the new feature) or should, if possible,
+   wait to be merged to `master` until after the release. Which path
+   is taken should be a collaboration between the release manager and
+   the developers of the change in question.
+5. New development unrelated to the new release can go directly into
+   `master` in parallel with a release stabilization.
+6. During stabilization, the release manager should update
+   `nixUnstable` to point to various commits of the `#-maintenance`
+   branch, to easily enable wider testing and integration.
+7. When the release manager determines readiness, the relevant commit
+   is announced to the community. Barring objections or last-minute
+   fixes judged valid by the release manager (or core team), the
+   commit is tagged and the branch goes into the maintenance phase.
+   Readiness is a judgment call, and should require increased
+   scrutiny/validation for a release with more complex/major changes.
+8. So long as no new feature work has yet happened on `master`,
+   `master` should be kept equal with the most recent `#-maintenance`
+   branch (through fast-forwards, not cherry-picks). Once new feature
+   work occurs, the semver minor (or major) versions are bumped and
+   normal development continues against `master`. This allows us to
+   maintain the "no patch release without a corresponding mainline
+   release" invariant (see point 12 below) while still allowing patch
+   releases to happen on the appropriate `#-maintenance` branch.
+9. Associated with each `NixOS` stable release is an associated
+   `maintained` series of `Nix`, with the same support lifetime as
+    the stable release.
+10. Any important bugfix can be cherry-picked from `master` to a
+    `#-maintenance` branch of a supported `maintained` series at any
+    committer's discretion.
+11. In rare cases, if a change truly only makes sense on a
+    `#-maintenance` branch (e.g. when we added support to `1.11` to
+    be able to work with a `2.0` database schema) it can be targeted
+    directly to the `#-maintenance` branch. This should be avoided if
+    possible and subject to especially careful testing and review.
+12. A maintenance release can be tagged off of the `#-maintenance`
+    branch at any time *after* there is a mainline release that
```

Comment:
Ah right, updated.

Reply:
Yes

Reply:
The idea is we never want the maintenance series to be "ahead" of the mainline series in any meaningful way.

Reply:
I understand the idea, but does that mean that «2.0.2» might be expected to wait for «2.1.0» release in case some feature development happens?

Reply:
Yes, that's the intention. That requires a certain level of stability on master, which I think is one of the goals here to encourage :grinning:

Reply:
Maybe define the mainline release explicitly then? Because «2.0.2» in my example could be mistaken for a mainline release (if there is no «2.1.0» or «3.0.0»).

Maybe just remove the word «mainline» and say something like «there is a release from a commit that diverges with the master branch after all of the master-branch versions of the fixes included in the desired maintenance release»?

Reply:
Does the updated point 3 help here?

Reply:
Not too much…

Reply:
Also, if there are 2.0.8, 2.1.6, 2.2.3 with 2.2.3 still being mainline, and a minor bug is found, does 2.0.9 normally require 2.1.7 to happen first/simultaneously or only 2.2.4?

Reply:
Currently we only have provisions for a single maintained series at any given time:

> Associated with each NixOS stable release is an associated maintained series of Nix, with the same support lifetime as the stable release.

But IMO if we ever do have more than one we should always update the newer ones first.

Reply:
I think .09 is still maintained-ish all April and .03 all October. At least for things that are annoying and simple to backport.